File: app_pojav_zh/src/main/java/com/movtery/pojavzh/ui/dialog/ProgressDialog.java
package com.movtery.pojavzh.ui.dialog;

import static net.kdt.pojavlaunch.prefs.LauncherPreferences.PREF_ANIMATION;

import android.content.Context;
import android.view.View;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;

import net.kdt.pojavlaunch.R;

public class ProgressDialog extends FullScreenDialog {
    private TextView textView;
    private ProgressBar progressBar;

    public ProgressDialog(@NonNull Context context, @NonNull OnCancelListener listener) {
        super(context);
        this.setContentView(R.layout.dialog_progress);
        this.setCancelable(false);

        init(listener);
    }

    private void init(OnCancelListener listener) {
        this.textView = findViewById(R.id.zh_download_upload_textView);
        this.progressBar = findViewById(R.id.progressBar2);
        Button cancelButton = findViewById(R.id.zh_download_cancel_button);

        this.progressBar.setMax(1000);
        cancelButton.setOnClickListener(v -> {
            if (listener != null) {
                if (!listener.onClick()) return;
            }
            dismiss();
        });
    }

    public void updateText(String text) {
        if (text != null) this.textView.setText(text);
    }

    public void updateProgress(double progress, double total) {
        double doubleValue = progress / total * 1000;
        int intValue = (int) (doubleValue);

        if (doubleValue > 0) this.progressBar.setVisibility(View.VISIBLE);
        else this.progressBar.setVisibility(View.GONE);
        this.progressBar.setProgress(intValue, PREF_ANIMATION);
    }

    public interface OnCancelListener {
        boolean onClick();
    }
}

File: app_pojav_zh/src/main/java/com/movtery/pojavzh/ui/dialog/ProgressDialog.kt
package com.movtery.pojavzh.ui.dialog

import android.content.Context
import android.view.View
import android.view.Window
import com.movtery.pojavzh.setting.AllSettings
import com.movtery.pojavzh.ui.dialog.DraggableDialog.DialogInitializationListener
import com.movtery.pojavzh.utils.file.FileTools.Companion.formatFileSize
import net.kdt.pojavlaunch.databinding.DialogProgressBinding

class ProgressDialog(context: Context, listener: OnCancelListener) : FullScreenDialog(context),
    DialogInitializationListener {
    private val binding = DialogProgressBinding.inflate(layoutInflater)

    init {
        this.setContentView(binding.root)
        this.setCancelable(false)
    }

    init {
        binding.progressBar.setMax(1000)
        binding.cancelButton.setOnClickListener {
            if (!listener.onClick()) return@setOnClickListener
            dismiss()
        }

        DraggableDialog.initDialog(this)
    }

    fun updateText(text: String?) {
        text?.apply { binding.textView.text = this }
    }

    fun updateRate(processingRate: Long) {
        if (processingRate > 0) binding.uploadRate.visibility = View.VISIBLE
        val formatFileSize = formatFileSize(processingRate)
        "$formatFileSize/s".also { binding.uploadRate.text = it }
    }

    fun updateProgress(progress: Double, total: Double) {
        val doubleValue = progress / total * 1000
        val intValue = doubleValue.toInt()

        binding.progressBar.apply {
            visibility = if (doubleValue > 0) View.VISIBLE else View.GONE
            setProgress(intValue, AllSettings.animation)
        }
    }

    override fun onInit(): Window? {
        return window
    }

    fun interface OnCancelListener {
        fun onClick(): Boolean
    }
}
